### PR TITLE
Migrate DesugaringOptimizer and Simplify expressions

### LINKED ISSUE #7146: Migrate DesugaringOptimizer to iterative optimizer

### LINKED ISSUE #7148: Migrate SimplifyExpressions to iterative optimizer

### PR DESCRIPTION
**Pull request by szymonm**
Resolves #7146 
Resolves #7148

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 1]
```diff
@@ -116,18 +116,6 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
...
-    public void initialize()
-    {
-        stats.export(exporter);
-    }
-
-    @PreDestroy
-    public void destroy()
-    {
-        stats.unexport(exporter);
-    }
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: is this change related?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExpressionRule.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+abstract class ExpressionRule
+        implements Rule
+{
+    abstract Expression rewrite(Expression expression, SymbolAllocator symbolAllocator, Session session);
+
+    @Override
+    public Pattern getPattern()
+    {
+        return Pattern.any();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExpressionRule.java]: I am sorry, this won't go ; /

You need to have separate rule per each node kind and have specialized pattern for them. That is why I was telling you about `RuleGroup` concept.

Please separate refactoring commits (those which does not change the functionality) from these which change the behavior of Presto

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DesugaringOptimizer.java [hunk 3]
```diff
@@ -1,211 +0,0 @@
...
-import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.getExpressionTypes;
-import static com.facebook.presto.sql.planner.ExpressionExtractor.extractExpressionsNonRecursive;
-import static com.google.common.base.Preconditions.checkState;
-import static com.google.common.collect.ImmutableList.toImmutableList;
-import static com.google.common.collect.ImmutableMap.toImmutableMap;
-import static java.util.Collections.emptyList;
-import static java.util.Objects.requireNonNull;
-
-@Deprecated
-public class DesugaringOptimizer
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DesugaringOptimizer.java]: Sorry, it has to stay for a while

**Comment by kokosing** [comment]
Additionally please add some tests

**Comment by kokosing** [comment]
Regarding refactoring I was thinking about something like that: https://gist.github.com/7ec809a63441bdb4a49505b82e8aa085 

What do you think?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 1 above)
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: by coincidence

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DesugaringOptimizer.java (hunk 3 above)
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DesugaringOptimizer.java]: Done.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 1 above)
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: This is actually only rearranging object methods after constructor (see after the constructor)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 4]
```diff
@@ -340,6 +328,18 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
...
+    public void initialize()
+    {
+        stats.export(exporter);
+    }
+
+    @PreDestroy
+    public void destroy()
+    {
+        stats.unexport(exporter);
+    }
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: I am not sure if we should touch this at all. They are special, so maybe they should be before the constructor.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionVerifier.java [hunk 5]
```diff
@@ -75,6 +75,25 @@
         this.symbolAliases = requireNonNull(symbolAliases, "symbolAliases is null");
     }
 
+    private static String getValueFromLiteral(Expression expression)
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionVerifier.java]: undo

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+public abstract class AbstractExpressionRewriteRuleSet
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java]: Please define a `RuleSet` interface

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+                new AggregationExpressionRewrite(),
+                new FilterExpressionRewrite(),
+                new TableScanExpressionRewrite(),
+                new JoinExpressionRewrite(),
+                new ValuesExpressionRewrite(),
+                new ApplyExpressionRewrite()
+        );
+    }
+
+    private static class Patterns
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java]: escalate it its own public class and define constants as methods like `join()` or `filter()`

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestAbstractExpressionRewriteRuleSet.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    {
+        @Override
+        Expression rewrite(Expression expression, Rule.Context context)
+        {
+            return new InPredicate(
+                    new LongLiteral("1"),
+                    new InListExpression(ImmutableList.of(
+                            new LongLiteral("1"),
+                            new LongLiteral("2")
+                    )));
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestAbstractExpressionRewriteRuleSet.java]: put this closing brackets to previous line, here in all other places.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        private static final Pattern PROJECT = Pattern.typeOf(ProjectNode.class);
+        private static final Pattern AGGREGATION = Pattern.typeOf(AggregationNode.class);
+        private static final Pattern FILTER = Pattern.typeOf(FilterNode.class);
+        private static final Pattern TABLE_SCAN = Pattern.typeOf(TableScanNode.class);
+        private static final Pattern JOIN = Pattern.typeOf(JoinNode.class);
+        private static final Pattern VALUES = Pattern.typeOf(ValuesNode.class);
+        private static final Pattern APPLY = Pattern.typeOf(ApplyNode.class);
+    }
+
+    final class ProjectExpressionRewrite
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java]: make them private

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java (hunk 6 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java]: Instead of making this abstract, pass a `Function<Expression, Expression>` to constructor

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+        {
+            return Patterns.PROJECT;
+        }
+
+        @Override
+        public Optional<PlanNode> apply(PlanNode node, Context context)
+        {
+            ProjectNode projectNode = (ProjectNode) node;
+            Assignments assignments = projectNode.getAssignments().rewrite(x -> AbstractExpressionRewriteRuleSet.this.rewrite(x, context));
+            return Optional.of(new ProjectNode(node.getId(), projectNode.getSource(), assignments));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java]: return `Optional.empty()` when there was no changes, here in all the other rules in this file

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 11]
```diff
@@ -151,7 +153,13 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
                         new PushProjectionThroughExchange()));
 
         builder.add(
-                new DesugaringOptimizer(metadata, sqlParser), // Clean up all the sugar in expressions, e.g. AtTimeZone, must be run before all the other optimizers
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: what happened with legacy rule?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarLambdaExpression.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+import com.facebook.presto.sql.tree.GroupingOperation;
+import com.facebook.presto.sql.tree.SymbolReference;
+
+import static com.google.common.base.Preconditions.checkState;
+
+public class DesugarLambdaExpression
+        extends AbstractExpressionRewriteRuleSet
+{
+    @Override
+    Expression rewrite(Expression expression, Rule.Context context)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarLambdaExpression.java]: add tests for this rewrite

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarAtTimeZone.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    private final SqlParser sqlParser;
+
+    public DesugarAtTimeZone(Metadata metadata, SqlParser sqlParser)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+    }
+
+    @Override
+    Expression rewrite(Expression expression, Rule.Context context)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarAtTimeZone.java]: add tests for this rewrite

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java [hunk 14]
```diff
@@ -80,7 +80,6 @@ public Pattern getPattern()
             Assignments assignments = projectNode.getAssignments().rewrite(x -> AbstractExpressionRewriteRuleSet.this.rewrite(x, context));
             return Optional.of(new ProjectNode(node.getId(), projectNode.getSource(), assignments));
         }
-
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java]: squash this commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java (hunk 6 above)
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java]: I don't see the point in introducing `interface RuleSet { Set<Rule> rules() }`. We are not going to use it anywhere for now. Can you explain why you think we need it?

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java]: For example there is ongoing work on migrating `PruneUnreferencedOutputs` and `PredicatePushdown`, where each rules looks very similar to each other, so I would expect them to take advantage of being able to be written next to each other, like you did for expressions rewrites.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 15]
```diff
@@ -151,7 +153,13 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
...
+                new IterativeOptimizer(
+                        stats,
+                        Sets.union(
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Please use `ImmutableSet.builder()` instead.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java (hunk 9 above)
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java]: Left them package private for testing. I can use `@VisibleForTesting` but then I need to define constructors explicitly. Do you prefer this way?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 15 above)
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: is there any other reason than consistency in this file?

**Comment by szymonm** [comment]
@kokosing it's ready to get back to

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarAtTimeZone.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+public class DesugarAtTimeZone
+        extends ExpressionRewriteRuleSet
+{
+    public DesugarAtTimeZone(Metadata metadata, SqlParser sqlParser)
+    {
+        super((expression, context) -> rewrite(expression, context, metadata, sqlParser));
+    }
+
+    static Expression rewrite(Expression expression, Rule.Context context, Metadata metadata, SqlParser sqlParser)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarAtTimeZone.java]: I would prefer not to copy paste code from legacy rules. Do the same thing as in other review, extract a legacy rewrite as some class and then use it here.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 17]
```diff
@@ -161,7 +164,14 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
                         new PushProjectionThroughExchange()));
 
         builder.add(
-                new DesugaringOptimizer(metadata, sqlParser), // Clean up all the sugar in expressions, e.g. AtTimeZone, must be run before all the other optimizers
+                new IterativeOptimizer(
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: I would glue this iterative optimizer with the subsequent. Ideally, plan is to have just a single iterative optimizer.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarLambdaExpression.java [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+        checkState(!(expression instanceof GroupingOperation), "GroupingOperation should have been re-written to a FunctionCall before execution");
+
+        if (expression instanceof SymbolReference) {
+            return expression;
+        }
+
+        expression = new LambdaCaptureDesugaringRewriter(context.getSymbolAllocator().getTypes(), context.getSymbolAllocator()).rewrite(expression);
+        return expression;
+    }
+}
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarLambdaExpression.java]: Please add some tests (for a expression rewrite).

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 19]
```diff
@@ -200,7 +199,11 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
                                         new PruneCountAggregationOverScalar()))
                                 .build()
                 ),
-                new SimplifyExpressions(metadata, sqlParser),
+                new IterativeOptimizer(
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Extract this as variable, and reuse.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyExpressions.java [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+import static java.util.stream.Collectors.toSet;
+
+public class SimplifyExpressions extends ExpressionRewriteRuleSet
+{
+    public SimplifyExpressions(Metadata metadata, SqlParser sqlParser)
+    {
+        super((expression, context) -> rewrite(expression, context, metadata, sqlParser));
+    }
+
+    static Expression rewrite(Expression expression, Rule.Context context, Metadata metadata, SqlParser sqlParser)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyExpressions.java]: The same comment as previously. Do not copy-paste.

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyExpressions.java]: I thought there were more rewrites.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java [hunk 21]
```diff
@@ -56,19 +58,28 @@
...
-            new ValuesExpressionRewrite(),
-            new ApplyExpressionRewrite());
+    private final ImmutableMap<Class<? extends PlanNode>, Rule<?>> allRules = ImmutableMap.<Class<? extends PlanNode>, Rule<?>>builder()
+            .put(ProjectNode.class, new ProjectExpressionRewrite())
+            .put(AggregationNode.class, new AggregationExpressionRewrite())
+            .put(FilterNode.class, new FilterExpressionRewrite())
+            .put(TableScanNode.class, new TableScanExpressionRewrite())
+            .put(JoinNode.class, new JoinExpressionRewrite())
+            .put(ValuesNode.class, new ValuesExpressionRewrite())
+            .put(ApplyNode.class, new ApplyExpressionRewrite())
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java]: maybe make the rewrite classes public and static, and make a constructor of `ERRS` to accept a set of rules

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java [hunk 22]
```diff
@@ -56,19 +58,28 @@
...
     public ExpressionRewriteRuleSet(ExpressionRewriteRule rewrite)
     {
         this.rewriter = rewrite;
+        this.rules = ImmutableSet.copyOf(allRules.values());
+    }
+
+    public ExpressionRewriteRuleSet(ExpressionRewriteRule rewrite, List<Class<? extends PlanNode>> supportedNodes)
+    {
+        this.rewriter = requireNonNull(rewrite);
+        this.rules = supportedNodes.stream().map(allRules::get).collect(toSet());
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java]:
```
 supportedNodes.stream()
    .map(allRules::get)
    .collect(toImmutableSet());
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 23]
```diff
@@ -21,6 +21,8 @@
 import com.facebook.presto.sql.planner.iterative.rule.AddIntermediateAggregations;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: IMO there is too many changes in single commit. But let it be this way.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarAtTimeZone.java [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+        if (expression instanceof SymbolReference) {
+            return expression;
+        }
+        Map<NodeRef<Expression>, Type> expressionTypes = getExpressionTypes(session, metadata, sqlParser, symbolAllocator.getTypes(), expression, emptyList()); // parameters already replaced
+
+        expression = ExpressionTreeRewriter.rewriteWith(new DesugaringRewriter(expressionTypes), expression);
+        return expression;
+    }
+
+    private static final List<Class<? extends PlanNode>> supportedNodes = ImmutableList.of(ProjectNode.class,
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarAtTimeZone.java]: each element in separate line, so ProjectNode should be in next line

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java (hunk 21 above)
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java]: I don't like the idea of ERRS constructor that takes a set of rules. ERRS is built around `ExpressionRewriteRule` and should use it. Adding a constructor of any Rule set seems like bypassing this concept and also enables creating ERRS that has nothing to do with rewriting expressions.

**Comment by szymonm** [comment]
Superseded by #8809 #8864
